File: packages/tui/src/form/builder/revert.ts
import { configurePrompts, promptEnvironment } from '#tui/environment';
import { Key } from '#tui/key';
import { renderError } from '#tui/theme';
import type { MaybePromise, PromptInput } from '#tui/types';

export class FormRevertedError extends Error {
	constructor() {
		super('Form reverted.');
		this.name = 'FormRevertedError';
	}
}

export const runWithFormRevert = async <T>(canRevert: boolean, callback: () => MaybePromise<T>): Promise<T> => {
	const environment = promptEnvironment();

	if (!environment.input.readKey) {
		return callback();
	}

	const previousInput = environment.input;

	const input: PromptInput = {
		...previousInput,
		async readKey(): Promise<string | null> {
			while (true) {
				const key = await previousInput.readKey?.();

				if (key !== Key.ctrlU) {
					return key ?? null;
				}

				if (canRevert) {
					throw new FormRevertedError();
				}

				environment.error.write(renderError('This cannot be reverted.'));
			}
		},
	};

	configurePrompts({ input });

	try {
		return await callback();
	} finally {
		configurePrompts({ input: previousInput });
	}
};

File: provision/go/internal/form/revert.go
package form

import (
	"errors"

	"github.com/oullin/tui/internal/environment"
	"github.com/oullin/tui/internal/key"
	"github.com/oullin/tui/internal/theme"
)

var ErrReverted = errors.New("form reverted")

type revertInput struct {
	canRevert bool
	previous  environment.Input
}

func runWithRevert(canRevert bool, callback func() (any, error)) (any, error) {
	env := environment.Current()
	if env.Input == nil {
		return callback()
	}

	return callbackWithInput(revertInput{canRevert: canRevert, previous: env.Input}, callback)
}

func callbackWithInput(input environment.Input, callback func() (any, error)) (any, error) {
	returnValue := any(nil)
	err := environment.With(environment.Patch{Input: input}, func() error {
		value, callbackErr := callback()
		returnValue = value
		return callbackErr
	})

	return returnValue, err
}

func (input revertInput) ReadKey() (string, bool, error) {
	pressed, ok, err := input.previous.ReadKey()
	if err != nil || !ok {
		return pressed, ok, err
	}

	if pressed != key.First(key.CtrlU) {
		return pressed, ok, nil
	}

	if input.canRevert {
		return "", false, ErrReverted
	}

	environment.Current().ErrorOutput.Write(theme.Red("  ⚠ This cannot be reverted.") + "\n")
	return input.ReadKey()
}

func (input revertInput) ReadLine(message string) (string, error) {
	return input.previous.ReadLine(message)
}
